zcash_client_sqlite/src/wallet/init/migrations/v_transactions_net.rs

```diff
@@ -0,0 +1,318 @@
+                           WHEN received_notes.memo IS NULL THEN 0
+                           ELSE 1
+                       END AS memo_present
+                FROM   transactions
+                       JOIN received_notes ON transactions.id_tx = received_notes.tx
+                UNION
+                SELECT transactions.id_tx            AS id_tx,
+                       transactions.block            AS mined_height,
+                       transactions.tx_index         AS tx_index,
+                       transactions.txid             AS txid,
+                       transactions.expiry_height    AS expiry_height,
+                       transactions.raw              AS raw,
+                       transactions.fee              AS fee,
+                       -sent_notes.value             AS value,
+                       CASE
+                           WHEN sent_notes.from_account = sent_notes.to_account THEN 0
+                           ELSE 1
+                       END AS sent_count,
+                       0                             AS is_change,
+                       0                             AS received_count,
+                       CASE
+                           WHEN sent_notes.memo IS NULL THEN 0
+                           ELSE 1
+                       END AS memo_present
+                FROM   transactions
+                       JOIN sent_notes ON transactions.id_tx = sent_notes.tx
+            ) AS notes
+            LEFT JOIN blocks ON notes.mined_height = blocks.height
+            GROUP BY notes.id_tx",
+        )?;
+
+        Ok(())
+    }
+
+    fn down(&self, _transaction: &rusqlite::Transaction) -> Result<(), WalletMigrationError> {
+        // TODO: something better than just panic?
+        panic!("Cannot revert this migration.");
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rusqlite::{self, params};
+    use tempfile::NamedTempFile;
+
+    use zcash_client_backend::keys::UnifiedSpendingKey;
+    use zcash_primitives::zip32::AccountId;
+
+    use crate::{
+        tests,
+        wallet::init::{
+            init_wallet_db, init_wallet_db_internal, migrations::add_transaction_views,
+        },
+        WalletDb,
+    };
+
+    #[test]
+    fn v_transactions_net() {
+        let data_file = NamedTempFile::new().unwrap();
+        let mut db_data = WalletDb::for_path(data_file.path(), tests::network()).unwrap();
+        init_wallet_db_internal(&mut db_data, None, &[add_transaction_views::MIGRATION_ID])
+            .unwrap();
+        let usk =
+            UnifiedSpendingKey::from_seed(&tests::network(), &[0u8; 32][..], AccountId::from(0))
+                .unwrap();
+        let ufvk = usk.to_unified_full_viewing_key();
+
+        db_data
+            .conn
+            .execute(
+                "INSERT INTO accounts (account, ufvk) VALUES (0, ?)",
+                params![ufvk.encode(&tests::network())],
+            )
+            .unwrap();
+
+        db_data.conn.execute_batch(
```

Comment:
```suggestion
        // - Tx 0 creates two notes of 2 and 5 zatoshis for account 0.
        // - Tx 1 creates two notes of 2 and 3 zatoshis for an external address, and a
        //   change note of 2 zatoshis.
        db_data.conn.execute_batch(
```
Writing this comment helped me clarify that this database is in a state that would never be produced by the existing wallet. Tx 1 is presumably intended to spend the two notes received in tx 0, but that requires setting the `spent` column for those two notes. Alternatively tx 1 represents a shielding transaction, but then there should be corresponding entries in the `utxos` table.

Reply:
`tx 0` simply represents a transaction received by the wallet from some external source; such inbound transactions can of course have multiple outputs to the same recipient.

Reply:
Sure, that part is fine; the issue was that tx 1's input state was not represented in the database. The fact that 7 zatoshis were received in tx 0 and then 7 zatoshis were spent in tx 1 led me to believe that it was tx 0 state that was missing from the test database state.